test/rgb_matrix/engine_test.exs

```diff
@@ -1,111 +1,175 @@
-  #     Frame.solid_color(pixels, Chameleon.Keyword.new("green")),
-  #     Frame.solid_color(pixels, Chameleon.Keyword.new("blue")),
-  #     Frame.solid_color(pixels, Chameleon.Keyword.new("white"))
-  #   ]
-
-  #   animation =
-  #     Animation.new(
-  #       type: Animation.Static,
-  #       frames: frames,
-  #       delay_ms: 10,
-  #       loop: 1
-  #     )
-
-  #   start_supervised!({Engine, {@leds, animation, [paintable]}})
-
-  #   Enum.each(frames, fn frame ->
-  #     assert_receive {:frame, ^frame}
-  #   end)
-  # end
-
-  # test "can play a different animation", %{paintable: paintable} do
-  #   {animation, _frame} = solid_animation()
-  #   {animation_2, frame_2} = solid_animation(127, 127, 127)
-
-  #   start_supervised!({Engine, {@leds, animation, [paintable]}})
-
-  #   :ok = Engine.play_animation(animation_2)
-
-  #   assert_receive {:frame, ^frame_2}
-  # end
-
-  # test "can register and unregister paintables", %{paintable: paintable} do
-  #   {animation, frame} = solid_animation()
-  #   {animation_2, frame_2} = solid_animation(127, 127, 127)
-
-  #   start_supervised!({Engine, {@leds, animation, []}})
-
-  #   :ok = Engine.register_paintable(paintable)
-
-  #   assert_receive {:frame, ^frame}
-
-  #   :ok = Engine.unregister_paintable(paintable)
-  #   :ok = Engine.play_animation(animation_2)
-
-  #   refute_receive {:frame, ^frame_2}
-  # end
+  use ExUnit.Case
+
+  alias RGBMatrix.{Animation, Engine}
+
+  setup_all [
+    :create_leds,
+    :create_frame,
+    :create_animation
+  ]
+
+  setup :create_paint_fn
+
+  test "can set an animation", %{
+    animation: animation
+  } do
+    assert Engine.set_animation(animation) == :ok
+  end
+
+  setup :set_animation
+
+  test "can register a paintable function", %{
+    paint_fn: paint_fn
+  } do
+    assert {:ok, ^paint_fn, _frame} = Engine.register_paintable(paint_fn)
+  end
+
+  test "renders frames", %{
+    frame: frame,
+    paint_fn: paint_fn
+  } do
+    {:ok, ^paint_fn, _frame} = Engine.register_paintable(paint_fn)
+    assert_receive {:frame, ^frame}
+  end
+
+  describe "interaction events" do
+    setup :create_interact_animation
+
+    test "call the interact/3 Animation callback", %{
+      interact_animation: interact_animation,
+      leds: leds,
+      paint_fn: paint_fn
+    } do
+      :ok = Engine.set_animation(interact_animation)
+      {:ok, ^paint_fn, _frame} = Engine.register_paintable(paint_fn)
+
+      interaction = hd(leds)
+      assert Engine.interact(interaction) == :ok
+      assert_receive {:interact, ^interaction}
+    end
+  end
+
+  describe "can unregister paintables" do
+    test "that have been registered", %{
+      frame: frame,
+      paint_fn: paint_fn
+    } do
+      {:ok, ^paint_fn, _frame} = Engine.register_paintable(paint_fn)
+
+      assert Engine.unregister_paintable(paint_fn) == :ok
+      assert maybe_receive_some_frames(frame, 6)
+    end
+
+    test "and will ignore invalid input", %{
```

Comment:
⚠️ (recommended)

I think the subtle difference here is not that `unregister_paintable` ignores invalid input, but rather that it's idempotent. So if you pass it `"NE!"` it's returning `:ok` because `"NE!"` is already "unregistered". For this reason I'd recommend changing the test name to reflect this, and unregistering `paint_fn` twice to demonstrate this.

Reply:
Yeah, I was thinking about that at some point, and I'm pretty sure it just slipped my mind.

I'll fix it up and get this merged in.